astropy/cosmology/flrw/base.py

```diff
@@ -194,14 +194,48 @@ def critical_density(self, z: Quantity | ArrayLike) -> Quantity:
         return self.critical_density0 * self.efunc(z) ** 2
 
 
+class _BaryonComponent:
+    """The cosmology has attributes and methods for the baryon density."""
+
+    Ob0: float
+    """Omega baryons: density of baryonic matter in units of the critical density at z=0."""
+
+    inv_efunc: Callable[[NDArray[Any]], NDArray[Any]]
```

Comment:
Should this be an abstract method?

Reply:
I don't think so. The best would probably be to have a `HasInvEFunc` protocol, but I think this suffices. It allows subclasses to implement `inv_efunc` however they want (method, callable attribute) so long as it's present for `Ob()` to call.

Reply:
> It is not obvious to me why both options should be supported

Nor to me, but the plan is factor out all these building blocks of a cosmology, organize them into a user-facing sub-module, build some of the classic pedagogical models, and make it easy for users to put these building blocks together to easily make custom cosmologies. So I'm keeping these mixin classes very general.

Reply:
We should keep in mind that expanding the public API is much simpler than restricting it, so it might be better to avoid making the API too flexible right now.

Reply:
True, but this is not an unreasonable scenario. In fact we kind of already do it in Astropy
https://github.com/astropy/astropy/blob/abc83b4e661d3abf720dfe4c68f04ae48c220258/astropy/cosmology/flrw/lambdacdm.py#L124C1-L125C82